kongswap_adaptor/src/deposit/test_failed_transfer_from.rs

```diff
@@ -0,0 +1,708 @@
+use super::*;
+use crate::kong_types::{AddTokenArgs, AddTokenReply, ICReply};
+use crate::tx_error_codes::TransactionErrorCodes;
+use crate::{
+    state::storage::ConfigState, validation::ValidatedTreasuryManagerInit, StableAuditTrail,
+    StableBalances, AUDIT_TRAIL_MEMORY_ID, BALANCES_MEMORY_ID,
+};
+use candid::Principal;
+use ic_stable_structures::memory_manager::MemoryManager;
+use ic_stable_structures::{Cell as StableCell, DefaultMemoryImpl, Vec as StableVec};
+use icrc_ledger_types::icrc1::transfer::{Memo, TransferArg};
+use kongswap_adaptor::agent::mock_agent::MockAgent;
+use maplit::btreemap;
+use pretty_assertions::assert_eq;
+use sns_treasury_manager::{
+    Allowance, Asset, Balance, BalanceBook, Balances, BalancesRequest, DepositRequest, Step,
+    TreasuryManager, TreasuryManagerInit, TreasuryManagerOperation,
+};
+use std::cell::RefCell;
+
+const E8: u64 = 100_000_000;
+
+use lazy_static::lazy_static;
+
+lazy_static! {
+    static ref SELF_CANISTER_ID: Principal =
+        Principal::from_text("jexlm-gaaaa-aaaar-qalmq-cai").unwrap();
+}
+
+fn make_approve_request(amount: u64, fee: u64) -> ApproveArgs {
+    ApproveArgs {
+        from_subaccount: None,
+        spender: Account {
+            owner: *KONG_BACKEND_CANISTER_ID,
+            subaccount: None,
+        },
+        // All approved tokens should be fully used up before the next deposit.
+        amount: Nat::from(amount - fee),
+        expected_allowance: Some(Nat::from(0u8)),
+        expires_at: Some(u64::MAX),
+        memo: None,
+        created_at_time: None,
+        fee: Some(fee.into()),
+    }
+}
+
+fn make_balance_request(self_id: Principal) -> Account {
```

Comment:
If this were a method on MockAgent, then it wouldn't need to take an explicit argument `self_id`